spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,197 @@
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.data.mongodb.MongoDbFactory;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.mongodb.core.query.Update;
+import org.springframework.integration.metadata.ConcurrentMetadataStore;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import com.mongodb.WriteResult;
+
+/**
+ * MongoDbMetadataStore implementation of {@link ConcurrentMetadataStore}.
+ * Use this {@link org.springframework.integration.metadata.MetadataStore} to
+ * achieve meta-data persistence shared across application instances and
+ * restarts.
+ *
+ * @author Senthil Arumugam, Samiraj Panneer Selvam
+ * @since 4.2
+ *
+ */
+public class MongoDbMetadataStore implements ConcurrentMetadataStore {
+
+	protected final Log logger = LogFactory.getLog(getClass());
+	private final MongoTemplate template;
+	private final String DEFAULT_COLLECTION_NAME = "metadatastore";
+	private final String collectionName;
+	private final String ID_FIELD = "_id";
+	private final String VALUE = "value";
+	
+	/**
+	 * 
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template The mongodb template
+	 * @param collectionName the collection name where it persists the data
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		this.template = template;
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template The mongodb template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
+		this(template, null);
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param factory The mongodb factory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory) {
+		this(factory, null);
+	}
+	
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * collection name
+	 * 
+	 * @param collectionName the collection name where it persists the data
+	 * @param factory MongoDbFactory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory, String collectionName) {
+		template = new MongoTemplate(factory);
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
+	
+	
+	/**
+	 * Method put.
+	 * @param key String
+	 * @param value String
+	 * @see org.springframework.integration.metadata.MetadataStore#put(String, String)
```

Comment:
Please, don't forget that Java 8 is strict for the JavaDocs. And describe, please, here which MongoDB operation performs this method.